### PR TITLE
agent/helm: Deprecate --kpr=partial|strict|disabled and use --kpr=true|false instead

### PR DESCRIPTION
**Pull request by brb**
See per commit.

Once the PR is merged, the following will be updated before the v1.14 release:

- [x] KPR GSG docs
- [ ] Upgrade guide
- [x] GH workflows update

### CODE REVIEW

File: daemon/cmd/kube_proxy_replacement_test.go [hunk 1]
```diff
@@ -185,11 +185,11 @@ func (s *KPRSuite) TestInitKubeProxyReplacementOptions(c *C) {
 			},
 		},
 
-		// KPR strict + no conntrack ipt rules: error, needs bpf masquerade
+		// KPR true + no conntrack ipt rules: error, needs bpf masquerade
 		{
 			"kpr-strict+no-conntrack-ipt-rules+no-bpf-masquerade",
```
Comment (sayboras) [inline daemon/cmd/kube_proxy_replacement_test.go]: nit: here as well?

File: daemon/cmd/daemon_main.go [hunk 2]
```diff
@@ -327,7 +327,7 @@ func initializeFlags() {
 	flags.String(option.EnablePolicy, option.DefaultEnforcement, "Enable policy enforcement")
 	option.BindEnv(Vp, option.EnablePolicy)
 
-	flags.Bool(option.EnableExternalIPs, defaults.EnableExternalIPs, fmt.Sprintf("Enable k8s service externalIPs feature (requires enabling %s)", option.EnableNodePort))
+	flags.Bool(option.EnableExternalIPs, false, fmt.Sprintf("Enable k8s service externalIPs feature (requires enabling %s)", option.EnableNodePort))
```
Comment (sayboras) [inline daemon/cmd/daemon_main.go:332]: do we need to mention upgrade note for these two flags ?

**Review by sayboras — APPROVED** [review]
LGTM, some nits as per below

**Review by jrajahalme — APPROVED** [review]

**Review by tklauser — APPROVED** [review]

**Review by aspsk — APPROVED** [review]

File: daemon/cmd/kube_proxy_replacement_test.go (hunk 1 above)
Comment (brb) [reply daemon/cmd/kube_proxy_replacement_test.go]: Good catch! Updated.

File: daemon/cmd/daemon_main.go (hunk 2 above)
Comment (brb) [reply daemon/cmd/daemon_main.go:332]: Aye, I am planning to do in the follow up PR to update the docs (I want to get this PR merged before the freeze :grin: )

**Comment by brb** [comment]
The ci-aks is notoriously flaky. While K8sUpstreamNetConformance is known to be broken (https://cilium.slack.com/archives/C7PE7V806/p1686843215310069).

Considering that @aanm was fine with PR, just wanted me to update the existing GHA workflows (done), and there is nothing for the github-sec team, this PR is OK with reviews.

File: Documentation/cmdref/cilium-agent.md [hunk 3]
```diff
@@ -227,7 +227,7 @@ cilium-agent [flags]
...
       --k8s-watcher-endpoint-selector string                    K8s endpoint watcher will watch for these k8s endpoints (default "metadata.name!=kube-scheduler,metadata.name!=kube-controller-manager,metadata.name!=etcd-operator,metadata.name!=gcp-controller-manager")
       --keep-config                                             When restoring state, keeps containers' configuration in place
-      --kube-proxy-replacement string                           Enable only selected features (will panic if any selected feature cannot be enabled) ("partial"), or enable all features (will panic if any feature cannot be enabled) ("strict"), or completely disable it (ignores any selected feature) ("disabled") (default "partial")
+      --kube-proxy-replacement string                           Enable only selected features (will panic if any selected feature cannot be enabled) ("false"), or enable all features (will panic if any feature cannot be enabled) ("true") (default "false")
```
Comment (nathanjsweet) [inline Documentation/cmdref/cilium-agent.md:231]: "true" and "false" don't really seem to be the right words to use. Why not just keep "partial" and "strict"?

**Review by nathanjsweet — CHANGES_REQUESTED** [review]
Everything looks good, but the "true", "false" options don't make semantic sense. "False" corresponds with the idea of "disabled" rather than "partial". Can we just keep the original words while eliminating the "disabled" option?

File: Documentation/cmdref/cilium-agent.md (hunk 3 above)
Comment (jrajahalme) [reply Documentation/cmdref/cilium-agent.md:231]: The problem is that the new `false` is not the same as the old `partial`. The change is that with false, host-port, external-ips, etc. need to be explicitly enabled, while with `partial` they were enabled by default.

Language-wise, `strict` and `partial` take much more explaining than `true` and `false`. Moreover, `false` here should be read to mean "full kube-proxy replecement has not been enabled"?

**Review by nbusseneau — APPROVED** [review]

**Review by nathanjsweet — APPROVED** [review]

**Comment by brb** [comment]
> Please triage the failures and highlight the issues that track the failures

https://github.com/cilium/cilium/pull/26036#issuecomment-1598232839

**Comment by joestringer** [comment]
That comment only alleges that certain things are flaky, but I don't see an issue to track it. How do we know what the failure is or whether it's related to this PR? How do we track who is working on it? If we just accept flaky CI and merge things, then those CI jobs will continue to complain and cause confusion for contributors. At that point, we may as well just delete the flaky workflows. 

At least taking 2m per issue to file the flake with the symptoms will help us to track frequency, identify whether the flake is affecting others, and have a chance at assigning & resolving the issues.

**Comment by joestringer** [comment]
I realize that the CI failures only apply to jobs that are not required, so that's a grey area. But also if we aren't tracking those failures and finding a volunteer to follow up, then maybe it's just a consistently broken workflow and it's providing zero value. Filing a dedicated issue makes it easier to bring up that specific issue in community meetings or slack channels and discuss what the solutions should be.

**Comment by joestringer** [comment]
Reviews are in, required CI jobs are passing. Merging.

**Comment by joestringer** [comment]
Why was this change made?

I'm hoping to follow up on the experience of having variable flags with "{disable,attempt-with-fallback,force-enable}" type semantics, which is effectively what we had before with "KPR={disable,partial,strict}". Given that for KPR we have moved away from that previous approach, what can we learn from the experience? I'm inferring from the docs update and the semantics in the end that we found the "enable if available" approach for "KPR=partial" was confusing and difficult to debug or assess how it is behaving. Is that a fair assessment?

**Comment by brb** [comment]
@joestringer 

> Why was this change made?

TL;DR It was causing **a lot** of confusion not only among end-users, but also engineers.

When the KPR feature was developed and the first versions of it released, many mainstream kernels did not fully support it. Therefore, we gave it a go for so-called feature detection (`probe`; default by the time KPR was implemented), which later got deprecated and removed. The removal was a good decision. Think of this nightmare - debugging a cluster, in which nodes run different kernels, thus different KPR feature set is enabled. We didn't have any CI coverage for such setups. Anyway, the upside of defaulting to `probe` was that it gave us confidence about stability of the features, as new users opted in.

Next, after removing `probe`, we were left with `partial | strict | disabled`. This was still causing lots of confusion. The `partial` was overlapping with `strict` and `disabled`. I.e., `partial` with all feature flags disabled (e.g., `--enable-node-port=false`) is the same as `disabled`. Meanwhile, `partial` with all feature flags enabled - `strict`. This realization lead to the deprecation and removal of `partial`.

Finally, with `strict | disabled` left, I started wondering - why not to change to more intuitive `true | false`, and then allow users to enable each feature when `KPR=false`. This would change the semantics of the `KPR` flag to "strictly enable all KPR features", which IMO is easier to grasp than all previous versions.

I think we all agree that less knobs means better for users and us.

> what can we learn from the experience?

Looking back, my learning from the whole KPR story is that we did not put enough thought into making it simple to use. Our expectation was that every user will follow the KPR GSG (which is gigantic by now). At the same time, we wanted (still want) to give a vast control to users to tune each feature (there are known limitations - e.g., the UDP and bpf_sock story). Having a table in the beginning of the GSG with what each feature/config does could have helped users (and us) to better understand it.
